core/src/validator.rs

```diff
@@ -749,6 +761,36 @@ fn active_vote_account_exists_in_bank(bank: &Arc<Bank>, vote_account: &Pubkey) -
     false
 }
 
+fn check_poh_speed(genesis_config: &GenesisConfig) {
+    if let Some(hashes_per_tick) = genesis_config.hashes_per_tick() {
+        let ticks_per_slot = genesis_config.ticks_per_slot();
+        let hashes_per_slot = hashes_per_tick * ticks_per_slot;
+        let hash_samples = 1_000_000;
+        let hash_time_ns = compute_hash_time_ns(hash_samples);
+        let my_ns_per_slot = (hash_time_ns * hashes_per_slot) / hash_samples;
+        debug!(
+            "computed: ns_per_slot: {} hash_time_ns: {}",
+            my_ns_per_slot, hash_time_ns
+        );
+        let target_ns_per_slot = genesis_config.ns_per_slot() as u64;
+        debug!(
+            "cluster ns_per_hash: {}ns ns_per_slot: {}",
+            target_ns_per_slot / hashes_per_slot,
+            target_ns_per_slot
+        );
+        if my_ns_per_slot < target_ns_per_slot {
+            let extra_ns = target_ns_per_slot - my_ns_per_slot;
+            info!("PoH speed check: Will sleep {}ns per slot.", extra_ns);
+        } else {
+            error!(
+                "Poh is slower than cluster target tick rate! mine: {} cluster: {}. If you wish to continue, try --ignore-poh-speed",
+                my_ns_per_slot, target_ns_per_slot,
+            );
+            abort();
```

Comment:
`std::process::exit(1)` would be a little less scary

Reply:
It does map to that:
```
fn abort() -> ! {
    #[cfg(not(test))]
    std::process::exit(1);

    #[cfg(test)]
    panic!("process::exit(1) is intercepted for friendly test failure...");
}
```

Reply:
Ah, got it!  I was under the assumption that was analogous to libc `abort()`